syntax = "proto3";
package zkp_auth;

/*
 * y1 = alpha^x mod p
 * y2 = beta^x mod p
 */ 
message RegisterRequest {
    string user = 1;
    bytes y1 = 2;
    bytes y2 = 3;
}

/*
 * Empty response for registration
 */
message RegisterResponse {}

/*
 * r1 = alpha^k mod p
 * r2 = beta^k mod p
 */ 
message AuthenticationChallengeRequest {
    string user = 1;
    bytes r1 = 2;
    bytes r2 = 3;
}

/*
 * auth_id is a unique identifier for the authentication session
 * c is the challenge sent back to the prover
 */
message AuthenticationChallengeResponse {
    string auth_id = 1;
    bytes c = 2;
}

/*
 * s = k - c * x mod q
 */ 
message AuthenticationAnswerRequest {
    string auth_id = 1;
    bytes s = 2;
}

/*
 * session_id is returned if the authentication is successful
 */
message AuthenticationAnswerResponse {
    string session_id = 1;
}

/* 
 * ZKP Authentication Service
 */
service Auth {
    /* 
     * Prover registers in the server sending y1 and y2
     */
     rpc Register(RegisterRequest) returns (RegisterResponse) {}
    /*
     * Prover ask for challenge in the server sending r1 and r2
     * Verifier sends the challenge "c" back
     */
    rpc CreateAuthenticationChallenge(AuthenticationChallengeRequest) returns (AuthenticationChallengeResponse) {}
    /*
     * Prover sends solution "s = k - c * x mod q" to the challenge
     * Verifier sends the session ID if the solution is correct
     */
    rpc VerifyAuthentication(AuthenticationAnswerRequest) returns (AuthenticationAnswerResponse) {}
}